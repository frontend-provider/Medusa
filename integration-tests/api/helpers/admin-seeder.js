const Scrypt = require("scrypt-kdf")
const { User } = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const buf = await Scrypt.kdf("secret_password", { logN: 1, r: 1, p: 1 })
  const password_hash = buf.toString("base64")

  await manager.insert(User, {
    id: "admin_user",
    email: "admin@medusa.js",
    api_token: "test_token",
    role: "admin",
    password_hash,
    ...data,
  })
}
