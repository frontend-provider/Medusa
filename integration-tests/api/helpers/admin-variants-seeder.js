const {
  Region,
  Product,
  ProductVariant,
  PriceList,
  CustomerGroup,
  Customer,
  Image,
  ShippingProfile,
  ProductCollection,
  ProductOption,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const yesterday = ((today) => new Date(today.setDate(today.getDate() - 1)))(
    new Date()
  )

  const tenDaysAgo = ((today) => new Date(today.setDate(today.getDate() - 10)))(
    new Date()
  )

  const defaultProfile = await manager.findOne(ShippingProfile, {
    where: { type: ShippingProfile.default },
  })

  const collection = manager.create(ProductCollection, {
    id: "test-collection",
    handle: "test-collection",
    title: "Test collection",
  })

  await manager.save(collection)

  await manager.insert(Region, {
    id: "reg-europe",
    name: "Test Region Europe",
    currency_code: "eur",
    tax_rate: 0,
  })

  await manager.insert(Region, {
    id: "reg-us",
    name: "Test Region US",
    currency_code: "usd",
    tax_rate: 0,
  })

  const customer = await manager.create(Customer, {
    id: "test-customer",
    email: "john@doe.com",
    first_name: "John",
    last_name: "Doe",
    password_hash:
      "c2NyeXB0AAEAAAABAAAAAVMdaddoGjwU1TafDLLlBKnOTQga7P2dbrfgf3fB+rCD/cJOMuGzAvRdKutbYkVpuJWTU39P7OpuWNkUVoEETOVLMJafbI8qs8Qx/7jMQXkN", // password matching "test"
    has_account: true,
  })

  const customerGroup = await manager.create(CustomerGroup, {
    id: "test-group",
    name: "test-group",
  })

  await manager.save(customerGroup)
  customer.groups = [customerGroup]
  await manager.save(customer)

  const priceListActive = await manager.create(PriceList, {
    id: "pl",
    name: "VIP sale",
    description: "All year sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  await manager.save(priceListActive)

  const priceListExpired = await manager.create(PriceList, {
    id: "pl_expired",
    name: "VIP summer sale",
    description: "Summer sale for VIP customers.",
    type: "sale",
    status: "active",
    starts_at: tenDaysAgo,
    ends_at: yesterday,
  })

  await manager.save(priceListExpired)

  const priceListWithCustomers = await manager.create(PriceList, {
    id: "pl_with_customers",
    name: "VIP winter sale",
    description: "Winter sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  priceListWithCustomers.customer_groups = [customerGroup]

  await manager.save(priceListWithCustomers)

  const productMultiReg = manager.create(Product, {
    id: "test-product",
    handle: "test-product-reg",
    title: "Multi Reg Test product",
    profile_id: defaultProfile.id,
    description: "test-product-description",
    status: "published",
    collection_id: "test-collection",
  })

  const image = manager.create(Image, {
    id: "test-image",
    url: "test-image.png",
  })

  productMultiReg.images = [image]

  await manager.save(productMultiReg)

  await manager.save(ProductOption, {
    id: "test-option",
    title: "test-option",
    product_id: "test-product",
  })

  const variantMultiReg = await manager.create(ProductVariant, {
    id: "test-variant",
    inventory_quantity: 10,
    title: "Test variant",
    variant_rank: 0,
    sku: "test-sku",
    ean: "test-ean",
    upc: "test-upc",
    barcode: "test-barcode",
    product_id: "test-product",
    prices: [
      {
        id: "test-price-multi-usd",
        currency_code: "usd",
        type: "default",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-usd",
        currency_code: "usd",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-expired-multi-usd",
        currency_code: "usd",
        amount: 70,
        price_list_id: "pl_expired",
      },
      {
        id: "test-price-multi-eur",
        currency_code: "eur",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-eur",
        currency_code: "eur",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-multi-eur-with-customer",
        currency_code: "eur",
        amount: 40,
        price_list_id: "pl_with_customers",
      },
      {
        id: "test-price-discount-expired-multi-eur",
        currency_code: "eur",
        amount: 70,
        price_list_id: "pl_expired",
      },
    ],
    options: [
      {
        id: "test-variant-option-reg",
        value: "Default variant",
        option_id: "test-option",
      },
    ],
  })

  await manager.save(variantMultiReg)
}
