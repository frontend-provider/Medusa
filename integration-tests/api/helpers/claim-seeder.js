const {
  ClaimOrder,
  Order,
  LineItem,
  Fulfillment,
  Return,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  let orderWithClaim = manager.create(Order, {
    id: "order-with-claim",
    customer_id: "test-customer",
    email: "test@email.com",
    payment_status: "captured",
    fulfillment_status: "fulfilled",
    billing_address: {
      id: "test-billing-address",
      first_name: "lebron",
    },
    shipping_address: {
      id: "test-shipping-address",
      first_name: "lebron",
      country_code: "us",
    },
    region_id: "test-region",
    currency_code: "usd",
    tax_rate: 0,
    discounts: [],
    payments: [
      {
        id: "test-payment-for-claim-order",
        amount: 10000,
        currency_code: "usd",
        amount_refunded: 0,
        provider_id: "test-pay",
        data: {},
      },
    ],
    items: [],
    ...data,
  })

  await manager.save(orderWithClaim)

  const li = manager.create(LineItem, {
    id: "test-item-co-2",
    fulfilled_quantity: 1,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 1,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li)

  const li2 = manager.create(LineItem, {
    id: "test-item-co-3",
    fulfilled_quantity: 4,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 4,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li2)

  const claimWithFulfillment = manager.create(ClaimOrder, {
    id: "claim-w-f",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  const ful1 = manager.create(Fulfillment, {
    id: "fulfillment-co-1",
    data: {},
    provider_id: "test-ful",
  })

  const ful2 = manager.create(Fulfillment, {
    id: "fulfillment-co-2",
    data: {},
    provider_id: "test-ful",
  })

  claimWithFulfillment.fulfillments = [ful1, ful2]

  await manager.save(claimWithFulfillment)

  const claimWithReturn = manager.create(ClaimOrder, {
    id: "claim-w-r",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  await manager.save(claimWithReturn)

  await manager.insert(Return, {
    id: "return-id-2",
    claim_order_id: "claim-w-r",
    status: "requested",
    refund_amount: 0,
    data: {},
  })
}
