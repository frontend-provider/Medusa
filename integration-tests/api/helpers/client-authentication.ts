const AUTH_COOKIE = {}
export async function getClientAuthenticationCookie(
  api,
  email = null,
  password = null
) {
  const user = {
    email: email ?? "test@medusajs.com",
    password: password ?? "test",
  }

  if (AUTH_COOKIE[user.email]) {
    return AUTH_COOKIE[user.email]
  }

  const authResponse = await api.post("/store/auth", user)
  AUTH_COOKIE[user.email] = authResponse.headers["set-cookie"][0].split(";")

  return AUTH_COOKIE[user.email]
}
