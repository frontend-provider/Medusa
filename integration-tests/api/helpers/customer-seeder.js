const { Customer, Address } = require("@medusajs/medusa")

module.exports = async (connection, data = {}) => {
  const manager = connection.manager

  await manager.insert(Customer, {
    id: "test-customer-1",
    email: "test1@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-2",
    email: "test2@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-3",
    email: "test3@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-has_account",
    email: "test4@email.com",
    has_account: true,
  })

  await manager.insert(Address, {
    id: "test-address",
    first_name: "Lebron",
    last_name: "James",
    customer_id: "test-customer-1",
  })
}
