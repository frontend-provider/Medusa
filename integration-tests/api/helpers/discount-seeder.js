const {
  ShippingProfile,
  Region,
  Discount,
  DiscountRule,
} = require("@medusajs/medusa")
module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  await manager.insert(Region, {
    id: "test-region",
    name: "Test Region",
    currency_code: "usd",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })

  await manager.insert(Region, {
    id: "test-region-2",
    name: "Test Region 2",
    currency_code: "eur",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })
}
