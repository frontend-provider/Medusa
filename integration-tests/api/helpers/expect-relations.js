const { expect } = require("@jest/globals")

export function expectRelations(expected, actual){
    expected.forEach(a => expectRelations_(a, actual))
}

function expectRelations_(expected, actual){
    const items = expected.split(".");
    let data_ = actual;

    for(const i in items){
        data_ = data_[items[i]];
        
        if(data_ instanceof Array && data_.length > 0) {
            data_ = data_[0];
        } 
    
        expect(data_).toBeDefined();
    }
}
