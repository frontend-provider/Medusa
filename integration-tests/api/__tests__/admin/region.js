const path = require("path")
const { Region } = require("@medusajs/medusa")

const setupServer = require("../../../helpers/setup-server")
const startServerWithEnvironment =
  require("../../../helpers/start-server-with-environment").default
const { useApi } = require("../../../helpers/use-api")
const { initDb, useDb } = require("../../../helpers/use-db")
const adminSeeder = require("../../helpers/admin-seeder")
const { simpleRegionFactory } = require("../../factories")

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

jest.setTimeout(30000)

describe("/admin/regions", () => {
  let medusaProcess
  let dbConnection
  const cwd = path.resolve(path.join(__dirname, "..", ".."))

  beforeAll(async () => {
    dbConnection = await initDb({ cwd })
    medusaProcess = await setupServer({ cwd })
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("Remove region from country on delete", () => {
    beforeEach(async () => {
      await adminSeeder(dbConnection)
      const manager = dbConnection.manager
      await manager.insert(Region, {
        id: "test-region",
        name: "Test Region",
        currency_code: "usd",
        tax_rate: 0,
      })

      await manager.query(
        `UPDATE "country" SET region_id='test-region' WHERE iso_2 = 'us'`
      )
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("successfully creates a region with countries from a previously deleted region", async () => {
      const api = useApi()

      const response = await api
        .delete(`/admin/regions/test-region`, {
          headers: {
            Authorization: "Bearer test_token",
          },
        })
        .catch((err) => {
          console.log(err)
        })

      expect(response.status).toEqual(200)
      expect(response.data).toMatchSnapshot({
        id: "test-region",
        object: "region",
        deleted: true,
      })

      const newReg = await api.post(
        `/admin/regions`,
        {
          name: "World",
          currency_code: "usd",
          tax_rate: 0,
          payment_providers: ["test-pay"],
          fulfillment_providers: ["test-ful"],
          countries: ["us"],
        },
        {
          headers: {
            Authorization: "Bearer test_token",
          },
        }
      )

      expect(newReg.status).toEqual(200)
      expect(newReg.data.region).toMatchSnapshot({
        id: expect.any(String),
        name: "World",
        currency_code: "usd",
        countries: [
          {
            region_id: expect.any(String),
          },
        ],
        tax_rate: 0,
        fulfillment_providers: [
          {
            id: "test-ful",
            is_installed: true,
          },
        ],
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })

  describe("GET /admin/regions", () => {
    beforeEach(async () => {
      const manager = dbConnection.manager
      await adminSeeder(dbConnection)
      await manager.insert(Region, {
        id: "test-region",
        name: "Test Region",
        currency_code: "usd",
        tax_rate: 0,
      })
      await manager.insert(Region, {
        id: "test-region-deleted",
        name: "Test Region",
        currency_code: "usd",
        tax_rate: 0,
        deleted_at: new Date(),
      })
      await manager.insert(Region, {
        id: "test-region-updated",
        name: "Test Region updated",
        currency_code: "usd",
        tax_rate: 0,
        updated_at: new Date(),
      })
      await manager.insert(Region, {
        id: "test-region-updated-1",
        name: "Test Region updated 1",
        currency_code: "usd",
        tax_rate: 0,
        updated_at: new Date("10/10/2000"),
      })
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("only returns non-deleted regions", async () => {
      const api = useApi()

      const response = await api
        .get(`/admin/regions`, {
          headers: {
            Authorization: "Bearer test_token",
          },
        })
        .catch((err) => {
          console.log(err)
        })

      expect(response.data.regions).toHaveLength(3)
      expect(response.data.regions).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: "test-region-updated-1",
          }),
          expect.objectContaining({
            id: "test-region",
          }),
          expect.objectContaining({
            id: "test-region-updated",
          }),
        ])
      )
      expect(response.status).toEqual(200)
    })

    it("filters correctly on update", async () => {
      const api = useApi()

      const response = await api
        .get(`/admin/regions?updated_at[gt]=10-10-2005`, {
          headers: {
            Authorization: "Bearer test_token",
          },
        })
        .catch((err) => {
          console.log(err)
        })

      expect(response.data.regions).toHaveLength(2)
      expect(response.data.regions).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: "test-region",
          }),
          expect.objectContaining({
            id: "test-region-updated",
          }),
        ])
      )
      expect(response.status).toEqual(200)
    })
  })

  describe("DELETE /admin/regions/:id", () => {
    beforeEach(async () => {
      await adminSeeder(dbConnection)
      const manager = dbConnection.manager
      await manager.insert(Region, {
        id: "test-region",
        name: "Test Region",
        currency_code: "usd",
        tax_rate: 0,
      })

      await manager.query(
        `UPDATE "country" SET region_id='test-region' WHERE iso_2 = 'us'`
      )
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("successfully deletes a region with a fulfillment provider", async () => {
      const api = useApi()

      // add fulfillment provider to the region
      await api.post(
        "/admin/regions/test-region",
        {
          fulfillment_providers: ["test-ful"],
        },
        {
          headers: {
            Authorization: "Bearer test_token",
          },
        }
      )

      const response = await api
        .delete(`/admin/regions/test-region`, {
          headers: {
            Authorization: "Bearer test_token",
          },
        })
        .catch((err) => {
          console.log(err)
        })

      expect(response.data).toEqual({
        id: "test-region",
        object: "region",
        deleted: true,
      })
      expect(response.status).toEqual(200)
    })

    it("fails to create when countries exists in different region", async () => {
      const api = useApi()

      try {
        await api.post(
          `/admin/regions`,
          {
            name: "World",
            currency_code: "usd",
            tax_rate: 0,
            payment_providers: ["test-pay"],
            fulfillment_providers: ["test-ful"],
            countries: ["us"],
          },
          {
            headers: {
              Authorization: "Bearer test_token",
            },
          }
        )
      } catch (error) {
        expect(error.response.status).toEqual(422)
        expect(error.response.data.message).toEqual(
          "United States already exists in region test-region"
        )
      }
    })
  })
})

describe("[MEDUSA_FF_TAX_INCLUSIVE_PRICING] /admin/regions", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    const [process, connection] = await startServerWithEnvironment({
      cwd,
      env: { MEDUSA_FF_TAX_INCLUSIVE_PRICING: true },
    })
    dbConnection = connection
    medusaProcess = process
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("POST /admin/regions/:id", () => {
    const region1TaxInclusiveId = "region-1-tax-inclusive"

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        await simpleRegionFactory(dbConnection, {
          id: region1TaxInclusiveId,
          countries: ["fr"],
        })
      } catch (err) {
        console.log(err)
        throw err
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should allow to create a region that includes tax", async function () {
      const api = useApi()

      const payload = {
        name: "region-including-taxes",
        currency_code: "usd",
        tax_rate: 0,
        payment_providers: ["test-pay"],
        fulfillment_providers: ["test-ful"],
        countries: ["us"],
        includes_tax: true,
      }

      let response = await api
        .post(`/admin/regions`, payload, adminReqConfig)
        .catch((err) => {
          console.log(err)
        })

      expect(response.data.region).toEqual(
        expect.objectContaining({
          id: expect.any(String),
          includes_tax: true,
          name: "region-including-taxes",
        })
      )
    })

    it("should allow to update a region that includes tax", async function () {
      const api = useApi()
      let response = await api
        .get(`/admin/regions/${region1TaxInclusiveId}`, adminReqConfig)
        .catch((err) => {
          console.log(err)
        })

      expect(response.data.region.includes_tax).toBe(false)

      response = await api
        .post(
          `/admin/regions/${region1TaxInclusiveId}`,
          {
            includes_tax: true,
          },
          adminReqConfig
        )
        .catch((err) => {
          console.log(err)
        })

      expect(response.data.region.includes_tax).toBe(true)
    })
  })
})
