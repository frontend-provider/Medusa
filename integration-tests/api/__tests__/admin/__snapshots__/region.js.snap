// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/admin/regions Remove region from country on delete successfully creates a region with countries from a previously deleted region 1`] = `
Object {
  "deleted": true,
  "id": "test-region",
  "object": "region",
}
`;

exports[`/admin/regions Remove region from country on delete successfully creates a region with countries from a previously deleted region 2`] = `
Object {
  "automatic_taxes": true,
  "countries": Array [
    Object {
      "display_name": "United States",
      "id": 236,
      "iso_2": "us",
      "iso_3": "usa",
      "name": "UNITED STATES",
      "num_code": 840,
      "region_id": Any<String>,
    },
  ],
  "created_at": Any<String>,
  "currency_code": "usd",
  "deleted_at": null,
  "fulfillment_providers": Array [
    Object {
      "id": "test-ful",
      "is_installed": true,
    },
  ],
  "gift_cards_taxable": true,
  "id": Any<String>,
  "metadata": null,
  "name": "World",
  "payment_providers": Array [
    Object {
      "id": "test-pay",
      "is_installed": true,
    },
  ],
  "tax_code": null,
  "tax_provider_id": null,
  "tax_rate": 0,
  "updated_at": Any<String>,
}
`;
